src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,148 @@
+            last.next = newNode;
+            newNode.previous = last;
+            last = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(value, null, null);
+        checkIndexStrict(index);
+        if (first == null) {
+            first = last = newNode;
+        } else if (index == 0) {
+            newNode.next = first;
+            first.previous = newNode;
+            first = newNode;
+        } else if (index == size) {
+            last.next = newNode;
+            newNode.previous = last;
+            last = newNode;
+        } else {
+            Node<T> previousElement = findNodeByIndex(index - 1);
+            newNode.next = previousElement.next;
+            newNode.previous = previousElement;
+            previousElement.next = newNode;
+            previousElement.next.previous = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-    }
-
-    @Override
-    public T get(int index) {
-        return null;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> setNewNode = findNodeByIndex(index);
+        T item = setNewNode.element;
+        setNewNode.element = value;
+        return item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> removedElement = findNodeByIndex(index);
+        T item = removedElement.element;
+        unlink(removedElement);
+        //size--;
+        return item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+
+        while (current != null) {
+            if (current.element == object || (object != null && object.equals(current.element))) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
+    @Override
+    public T get(int index) {
+        checkIndex(index);
+        Node<T> node = findNodeByIndex(index);
+        return node.element;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexStrict(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index of bounds: ");
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index of bounds: ");
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> current = first;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
```

Comment:
[Don’t iterate your nodes from the head to tail or vice versa.](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-iterate-your-nodes-from-the-head-to-tail-or-vice-versa)